### ISSUE TITLE
Support the `pow` method for `Decimal`/floating point arguments

### ISSUE THREAD

**Issue by AdminXVII** [paupino/rust-decimal#169]
Ion would like to remove its dependency to C, and the next step would be to remove the `decimal` crate. Unfortunately, not being able to raise some Decimal to a floating point power is a serious barrier to our objective.

**Comment by paupino** [comment]
I'll take a look into how to efficiently implement this!

**Comment by paupino** [comment]
Side note, it'll probably be leveraging https://docs.rs/num/0.2.0/num/pow/trait.Pow.html

**Comment by ambaxter** [comment]
Adding to that - logn, log10, ln, sqrt, and nth_root would be really nice :)

**Comment by paupino** [comment]
Funnily enough, in order to implement `pow` completely I need `ln` support. Consequently, this is partially in scope!

**Comment by c410-f3r** [comment]
Any update regarding this matter?

**Comment by paupino** [comment]
I'm afraid I've been a little side tracked lately - I'll see if I can kickstart this again since we're going into holidays and give an updated timeline.

**Comment by c410-f3r** [comment]
I could do my own research and implement a high performance `pow`, `ln` or `log` implementation but it would probably take some time because of my inexperience at this field.
Having said that, can I start a bounty for this issue?

**Comment by paupino** [comment]
Certainly welcome to. Implementing `pow` for integers is relatively easy, however for decimals requires implementing logarithmic tables. 

Out of curiosity, do you need this for decimal based power functions? Also, would conversion to `f64` and then back work?

**Comment by AdminXVII** [comment]
Conversion to f64 could work if that makes it easier for you. The application is a scientific calculator, so while it could be useful to have exact results, f64 approximation would be ok I believe.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 91.0 TUSD (91.0 USD @ $1.0/TUSD)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paupino/rust-decimal/169/4456).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/paupino/rust-decimal/169/4456).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://chat.gitcoin.co/'>Gitcoin Chat</a>
* $129,138.71 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by c410-f3r** [comment]
It took me more than 6 months to gather $100. Jokes aside, it isn't much but is what I could afford to spend now and the amount might be increased in the future.

For those that wanna try, the work must contain unit tests to assert logic and a benchmark against native arithmetic operations to measure the implementation quality.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 3 months from now.
Please review their action plans below:


**1) [hellomalaysia](https://gitcoin.co/hellomalaysia) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/paupino/rust-decimal/169/4456?mutate_worker_action=approve&worker=hellomalaysia) | [reject worker](https://gitcoin.co/issue/paupino/rust-decimal/169/4456?mutate_worker_action=reject&worker=hellomalaysia))_.**

understand that this is an application and I should wait for approval from the funder before starting work.
I agree to keep the funder informed of my progress every few days.

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/paupino/rust-decimal/169/4456).

**Comment by kovkev** [comment]
c410-f3r, shouldn't you provide the tests that need to be fulfilled?

**Comment by c410-f3r** [comment]
OK, I abused the term "arithmetic". The tests might contain some arithmetic operation but in general they should assert that the correct behavior of pow. E.g.:

```
assert_eq!(Decimal::one().pow(0), Decimal::zero());
assert_eq!(Decimal::zero().pow(1), Decimal::zero());

// Rustc outputs 1 for 0^0 and I am not sure if the following should panic
assert_eq!(Decimal::zero().pow(0), Decimal::one());

... the list goes on
```

**Comment by hellomalaysia** [comment]
Done

**Comment by hellomalaysia** [comment]
https://github.com/paupino/rust-decimal/issues/169#issuecomment-645724684

> OK, I abused the term "arithmetic". The tests might contain some arithmetic operation but in general they should assert that the correct behavior of pow. E.g.:
> 
> ```rust
> assert_eq!(Decimal::one().pow(0), Decimal::one());
> assert_eq!(Decimal::zero().pow(1), Decimal::zero());
> 
> // Rustc outputs 1 for 0^0 and I am not sure if the following should panic
> assert_eq!(Decimal::zero().pow(0), Decimal::one());
> 
> ... the list goes on
> ```

**Comment by buddyhill** [comment]
:]

**Comment by vineetred** [comment]
Hey, is the issue/bounty still accepting people?

**Comment by c410-f3r** [comment]
@vineetred Yeah. Good luck!

**Comment by vineetred** [comment]
@c410-f3r Just to be clear, I need to build a method in Rust that can raise floats to decimal powers. Is this right?

**Comment by c410-f3r** [comment]
Not sure if the exponential part should be a float, an integer or another decimal. Maybe (decimal base)^(decimal exponential) = decimal result? @paupino What do you think about it?

Heh, I purposefully created this bounty to avoid getting into algorithm details. Other than that, feel free to contact me; my email is in my GitHub profile.

**Comment by jbg** [comment]
surely not a float type, neither as the base nor the exponent. `Decimal` for both would make sense, to me at least

**Comment by paupino** [comment]
I agree with @jbg here - I think it should be a `Decimal` for both the exponent and the base. 

Just for further info: a while back I did try some experimentation with this ticket on [this branch](https://github.com/paupino/rust-decimal/compare/feature/pow). I think that the implementation will need to branch when the exponent scale is equal to zero and is greater than zero. The equal to zero scenario (i.e. an integer exponent) is relatively straight forward however as soon as you start using a scale greater than zero (i.e. a fractional exponent) you start needing to look into other methods such as log tables. This has a pretty good breakdown it seems: https://medium.com/coinmonks/math-in-solidity-part-5-exponent-and-logarithm-9aef8515136e

Anyway, long story short I'd suggest `Decimal` for both - I actually think it'll make things more efficient in the long run as the binary format is very explicit.

**Comment by KiChjang** [comment]
@paupino I pulled from your branch and started working on fixing some unit tests, and now I'm onto the part where we have decimal exponents. I just realized that any fractional exponents can be rewritten as taking the nth root of the base, but here's what the interesting part is -- when the base is negative, the nth root exists only if n is odd, otherwise taking the nth root produces an imaginary number.

I currently panic when the base is negative, regardless of whether n is even or odd. Do you think the problem I raised should be solved by this issue, or could it be an enhancement for the future?

**Comment by paupino** [comment]
@KiChjang I'd probably extend this question to @c410-f3r for requirements here. 
If it's not needed for the bounty then I'd be happy to create another ticket to implement this functionality separately, so long as the code was explicit such that it either returned an error stating that it's unsupported or panicked similarly (i.e. a "handled" panic).

**Comment by c410-f3r** [comment]
Me? I am but a simple peasant!

Just kidding. Maybe do what std does?

```rust
assert_eq!(-64.0f64.sqrt(), -8.0);
```

As you can see, it just returns the negative number of a normal square root. Not sure about n^(-3, -4, -5, ..., -m) or what other decimal C/C++ libraries do in this case.

It is also possible to include alternative versions of `pow`, e.g.:

```rust
// Just options:

// Optioned algorithm that makes inputs assumptions and always returns valid values
pow(n: Decimal) -> Decimal;

// Returns None if n is negative
pow_opt(n: Decimal) -> Option<Decimal>;

// Panics at run-time if n is negative
pow_checked(n: Decimal) -> Option<Decimal>;
```

Personally, the most important thing to me is performance. What good is an algorithm that has poor performance when it is currently possible to perform `Decimal -> f64 -> pow -> Decimal` in a reasonable time?

**Comment by KiChjang** [comment]
@c410-f3r Your code example is a bit erroneous because what actually happens is that NaN results if you square rooted it properly. `-64f64.sqrt()` is interpreted as `-(64f64.sqrt())`, so when I did `(-64f64).sqrt()`, it gave me `NaN`.

Anyhow, it looks to me that we should really be panicking for now, and only worry about negative bases in the future.

**Comment by c410-f3r** [comment]
Oopppssss, didn't notice that. What a shame :(

**Comment by paupino** [comment]
@KiChjang There is no NaN representation at the moment. You could get equivalent information back by returning an enum if you really wanted to.

**Comment by c410-f3r** [comment]
Looks like no-one was able to complete the challenge in these 3 months :/

The bounty is canceled for the time being but I will be back in the future with my own PR or a increased bounty value.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__The funding of 91.0 TUSD  (91.0 USD @ $1.0/TUSD) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://chat.gitcoin.co/'>Gitcoin Chat</a>
* $760,605.42 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by vitcra** [comment]
I am curious about the back story that lead to choosing the particular representation for Decimal used here. The fact that it does not adhere to the IEEE-754 standard, in my opinion, isolates it from much of the scientific literature out there and more importantly the guarantees it provides for the correctness and error estimates of the operations needed to implement pow().

**Comment by jbg** [comment]
> I am curious about the back story that lead to choosing the particular representation for Decimal used here. The fact that it does not adhere to the IEEE-754 standard, in my opinion, isolates it from much of the scientific literature out there and more importantly the guarantees it provides for the correctness and error estimates of the operations needed to implement pow().

Does IEEE754 standardise some fixed-point decimal representation? I am only familiar with its floating point representations which don't seem relevant for implementing a fixed-point decimal type.

**Comment by vitcra** [comment]
I see, no, you are right, it does not standardise fixed-point representation. Was to focused on the initial need to eventually replace `decimal` crate which implements floating point arithmetic.

**Comment by paupino** [comment]
Since `powi` is supported, I'm going to rename this issue to represent supporting fixed precision arguments.

**Comment by c410-f3r** [comment]
Thank you very much @paupino!
